File: src/slug.ts
import { Err, Ok, type Result } from '@atomicloud/diene.result';

/**
 * Normalize an arbitrary string into a deterministic kebab-case slug.
 *
 * The transform is: NFKD normalization, combining-mark removal, lowercasing,
 * trimming, and collapsing every run of non-alphanumeric characters into a
 * single hyphen (with leading/trailing hyphens stripped).
 */
export function slugify(input: string): string {
  return input
    .normalize('NFKD')
    .replace(/[\u0300-\u036f]/g, '')
    .toLowerCase()
    .trim()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '');
}

/** Validation failure raised when a namespace or key slugifies to empty. */
export class NamespacedKeyValidationError extends Error {
  constructor(
    readonly field: 'namespace' | 'key',
    readonly reason: string,
  ) {
    super(`${field} ${reason}`);
    this.name = 'NamespacedKeyValidationError';
  }
}

/**
 * Compose a `namespace:key` identifier from slugified parts.
 *
 * This is a total function: it never throws for validation failures. When a
 * part slugifies to empty it returns an `Err` carrying a
 * {@link NamespacedKeyValidationError}; otherwise it returns an `Ok` with the
 * composed identifier.
 */
export function namespacedKey(namespace: string, key: string): Result<string, NamespacedKeyValidationError> {
  const ns = slugify(namespace);
  if (ns === '') {
    return Err<string, NamespacedKeyValidationError>(
      new NamespacedKeyValidationError('namespace', 'must not be empty'),
    );
  }

  const k = slugify(key);
  if (k === '') {
    return Err<string, NamespacedKeyValidationError>(new NamespacedKeyValidationError('key', 'must not be empty'));
  }

  return Ok<string, NamespacedKeyValidationError>(`${ns}:${k}`);
}

File: Lib/Slug.cs
using System.Globalization;
using System.Text;

namespace AtomiCloud.Diene.CoreUtils;

/// <summary>
/// Deterministic kebab-case identifiers. The transform is byte-parity with the
/// sibling <c>@atomicloud/diene.core-utils</c> <c>slugify</c>, so the same input
/// yields the same identifier in every language of the family.
/// </summary>
public static class Slug
{
    /// <summary>
    /// Folds an arbitrary string into a deterministic kebab-case slug: NFKD
    /// normalization, combining-mark removal, lowercasing, then every run of
    /// non-alphanumeric characters collapsed to a single hyphen with leading and
    /// trailing hyphens stripped.
    /// </summary>
    public static string Slugify(string input)
    {
        ArgumentNullException.ThrowIfNull(input);

        var folded = new StringBuilder(input.Length);
        foreach (var character in input.Normalize(NormalizationForm.FormKD))
        {
            if (CharUnicodeInfo.GetUnicodeCategory(character) != UnicodeCategory.NonSpacingMark)
                folded.Append(char.ToLowerInvariant(character));
        }

        var slug = new StringBuilder(folded.Length);
        var pendingSeparator = false;
        foreach (var character in folded.ToString())
        {
            if (char.IsAsciiLetterLower(character) || char.IsAsciiDigit(character))
            {
                if (pendingSeparator && slug.Length > 0) slug.Append('-');
                pendingSeparator = false;
                slug.Append(character);
            }
            else
            {
                pendingSeparator = true;
            }
        }

        return slug.ToString();
    }

    /// <summary>
    /// Composes a <c>namespace:key</c> identifier from slugified parts. This is a
    /// total function: a part that slugifies to empty yields an error instead of
    /// an exception.
    /// </summary>
    public static Result<string, KeyError> NamespacedKey(string ns, string key)
    {
        ArgumentNullException.ThrowIfNull(ns);
        ArgumentNullException.ThrowIfNull(key);

        var slugNamespace = Slugify(ns);
        if (slugNamespace.Length == 0)
            return Result.Err<string, KeyError>(new KeyError("namespace must not be empty", ns));

        var slugKey = Slugify(key);
        if (slugKey.Length == 0)
            return Result.Err<string, KeyError>(new KeyError("key must not be empty", key));

        return Result.Ok<string, KeyError>($"{slugNamespace}:{slugKey}");
    }
}
